[WIP] Landmark based shuttles

mwerezak opened the pull request:
Shuttles that use landmarks to specify destinations instead of areas

DNM for now.

This is an idea that's been on my mind for a while. Recently I noticed that thanks to a number of past refactoring of move_contents_to(), we're actually not all that far away from being able to do it, at least for ferry shuttles.

 The overmap shuttles will be a bit more work, which I didn't fully anticipate. It's still doable though, I think.

The main advantage is that it open up the possibility of sending shuttles anywhere without needing prepainted destination areas. Instead, creating a new shuttle destination is just a matter of plopping down a landmark, which I think is valuable.

There will not be collision checking in this PR, but that can be done sometime down the road. I guess the other thing is that the shuttle will have a single area that defines it's shape, which will move with the shuttle. I don't anticipate that being a problem, however I'm not really familiar with the details of working with areas, so hopefully there's no hidden gotchas with that.

mwerezak committed:
Removes unused direction parameter, enables check_solid by default

mwerezak committed:
Refactor move_contents_to()

Refactors move_contents_to() so that groups of turfs can
be translated by specifying a source and destination origin
Also moved the procs from unsorted.dm to turfs.dm helpers

mwerezak committed:
Shuttle areas -> shuttle landmarks

Removed prototypes from parameters to shuttle methods,
since the argumenst are no longer always areas.
Makes a shuttle's area move with the shuttle.
Updates shuttle creating to locate landmarks.

mwerezak committed:
Updates overmap shuttle

mwerezak commented:
Updated PR description.

BlueNexus commented:
Welcome back, Harpy!

mwerezak commented:
Thanks!

mwerezak committed:
Makes navpoints separate from landmarks

Fixes them deleting themselves...
note to self: don't inherit from bizarre legacy types

mwerezak committed:
Updates "overmap_example" maps and map-code

mwerezak committed:
Ferry shuttle refactor

mwerezak committed:
Rename /obj/effect/shuttle_nav -> shuttle_landmark

mwerezak committed:
Refactor shuttle base type

Shuttles now track their current area and landmark.
Shuttle move/launch procs no longer take an origin location as argument,
there was never a valid reason to supply anything other than their current
area/landmark, so they always just use that now instead.

mwerezak commented:
I think I may have not merged this correctly...

Kasuobes commented:
Revert revert revert!

mwerezak commented:
My git-fu is stronger now

mwerezak committed:
Updates dme

mwerezak committed:
Adds example map ferry shuttle

mwerezak committed:
Fix arguments not being forwarded to parent constructor

mwerezak committed:
Fixed unsimulated atoms transporting at map edge

mwerezak committed:
Expands example map

So that example shuttle are no longer inside the 7-tile edge boundary

mwerezak committed:
Adds example map shuttle docking arms

mwerezak committed:
Updates autodock shuttles (and subtypes) to handle docking using waypoints

PsiOmegaDelta committed:
Merge remote-tracking branch 'upstream/dev' into landmark-shuttles

mwerezak committed:
Fixes door frequencies on example map, space under doors

m2

mwerezak committed:
Updates global var access

mwerezak committed:
Removes waypoint datum, waypoint repository

mwerezak committed:
Finishes up overmap example map changes

mwerezak commented:
Cleaned up a lot of the inheritance spaghetti that made shuttle code a little hard to follow sometimes. Now the base shuttle only cares about what it needs to use to move shuttles around, no more proc litter for stuff that is only used by subtypes. Docking related behaviour is now handled largely by the new autodock subtype. 

Docking now works with shuttles in both the example and example overmap maps (tested too!). Just need to reimplement shuttles throwing atoms and gibbing mobs, shuttle ceilings  (largely unchanged, just a different area is used), collision/free space checks (planning to do simple area checking for this PR), then update the unit tests, and this will be code complete.



Then it's on to the mapping work to update exodus and torch. The commits will probably need some squashing after that as well.

mwerezak commented:
Oh, well, can't forget multi shuttles. But those should be straightforward to implement on top of autodock. Can probably reuse the overmap shuttle nanoui template for it, too.

mwerezak committed:
Fixes docking controller initialization race condition

mwerezak committed:
Updates multi shuttles

atlantiscze commented:
I'm currently working on a PR that's also highly shuttle-related (bluespace drives), how's this progressing? I'd prefer to avoid merge conflicts.

mwerezak commented:
It's more since than not-done at this point. Probably around 70% offhand? It will probably be done next weekend, if not then the weekend after.

mwerezak committed:
Merge branch 'dev' of https://github.com/Baystation12/Baystation12 into landmark-shuttles

mwerezak committed:
shuttle/current_location is now /obj/effect/shuttle_landmark

Removes current_waypoint var, renames next_waypoint -> next_location

mwerezak committed:
Fixes transported mobs having incorrect lastarea

Changing a turfs area doesn't appear to update lastarea,
so we update the areas before moving mobs.

mwerezak commented:
Okay, the last of the stuff in the way of map changes should finally be clear.

Unfortunately I've had some strange issues squashing the previous commits, squashing seems to result in massive amounts of merge conflicts even when not squashing across merge nodes and when squashing only after the line-ending changes. So... may not be able to get this in under 27-something commits.

mwerezak commented:
Guess map stuff will wait until next weekend, most likely.

PsiOmegaDelta commented:
Squashing while having merge commits does have a tendency to not work out very well.

mwerezak committed:
Collision checks, admin verbs

mwerezak committed:
Updates escape pods

mwerezak committed:
Merge branch 'dev' of https://github.com/Baystation12/Baystation12 into landmark-shuttles

mwerezak committed:
Simplifies mapping docking controller tags

comma commented:
Feel free to reopen when you're back from the dead!
